### ISSUE TITLE
New messages arrive only after I've opened the app and unlocked the database

### ISSUE THREAD

**Issue by WalterFaber** [WhisperSystems/TextSecure#2852]
Since the update yesterday new messages arrive only after I've opened the app and unlocked the database.

Prior to the update, I received new messages even if I hadn't opened the app for hours (database is locked after 15 Minutes)

I use version 2.9.3 and android 4.1.1

Kind Regards
Walter

**Comment by manhole11** [comment]
I've had the same problem. However, I don't have a passphrase timeout. I believe that the app crashes and doesn't lock because of the timeout... we should get a debug log

**Comment by moxie0** [comment]
@WalterFaber Does this happen for every message, or just some messages?

**Comment by moxie0** [comment]
Also have you ever used a backup/restore tool like Tintanium backup, have you recently switched phones, or done anything at all unusual?

**Comment by meryl1** [comment]
I can confirm the issue, I keep opening the app to make sure I'm receiving messages...

**Comment by WalterFaber** [comment]
Yes, it happened to every message as long as Textsecure is not open and active. I just received a message correctly while Textsecure was running.

Nothing elese changed - until I made the update yesterday, it had been working.

I also rebooted my phone 2 times in the meantime. That doesn't help.

**Comment by moxie0** [comment]
@WalterFaber Thanks, if you're comfortable doing it, would you mind emailing me your phone number so that i can check and see if the server has a gcm id for you? moxie.github@whispersystems.org

**Comment by WalterFaber** [comment]
Yes I can do that... did you happen to have a s/mime certificate?

**Comment by moxie0** [comment]
@WalterFaber I do not.

**Comment by WalterFaber** [comment]
Done :-) -> ABC

**Comment by moxie0** [comment]
Hmm, everything looks fine. Did you disable "background data" for google services, or anything like that?  Do you use stock android, or an aftermarket rom with a gapps install?

**Comment by WalterFaber** [comment]
No I didn't change any config since yesterday (actually, I had the issue with the background data when I started to use TS several months ago, but I fixed it at that time)

I have a HTC phone with the standard "htc sense" layer.

In my e-mail inbox I found messages from two other users who claimed to have the same problem. They look like replies to this conversation but I can't see them here. Can you see them?

Is there a way to send myself test-messages?

**Comment by moxie0** [comment]
Let's get a debug log then.  Try this:

1) Put the app into the background for > 1min with no incoming traffic.
2) Send yourself a message from another device
3) Wait a few seconds
4) When you don't get the message, open the app.  The message should appear.  At  that point capture a debug log from the TextSecure Settings -> Advanced -> Submit debug log.  Post that link here.

If you don't have access to another device, I can send you a message.  Just post here when you'd like me to.

**Comment by WalterFaber** [comment]
Ok... are there any confidential information in the log (passwords, keys, telephone number, name, e-mail address)? IPs are ok.

Just send me a message in 2 minutes. I will keep an eye on my phone

**Comment by moxie0** [comment]
There shouldn't be, just the last two digits of numbers you send messages to during the log's period are printed, the rest are removed.

**Comment by moxie0** [comment]
message sent, you can go ahead and open the app to capture the debug log now

**Comment by WalterFaber** [comment]
massage arrived after I opened the app... so the error occured... give me second please

**Comment by moxie0** [comment]
@WalterFaber It's important to have captured the debug log *immediately* after opening the app and receiving the message.

**Comment by WalterFaber** [comment]
argh.. is it already too late? I'm ready now to submit it

**Comment by moxie0** [comment]
@WalterFaber Yes, 10 minutes is too long. Let's start over.

**Comment by WalterFaber** [comment]
sorry... app is in the background... please send the message in 2 minutes

**Comment by moxie0** [comment]
Message sent, you can open the app.

**Comment by WalterFaber** [comment]
ok... the error was reproduceable ... here's the log 
https://gist.github.com/c6783f8eefd60a9d4222

**Comment by moxie0** [comment]
Hmm, from the log it's clear that you're not receiving any GCM messages at all, but I can confirm that the server is sending them to you.  That's pretty unusual for a non-rooted device.

Was there anything strange about the moment where the app updated?

**Comment by WalterFaber** [comment]
No, nothing strange. I updated it as I always do... could you send me another test-message please?
I had a background data limitation for google play store... I don't know when this was set, but I haven't changed that for months... so it shouldn't affect the gcm messages... but just to make sure (this was the only limitation)

**Comment by moxie0** [comment]
Just sent another message.  I think background data for Play does actually effect GCM. If you could turn off all background data restrictions, it'd be good just to rule that out.

**Comment by WalterFaber** [comment]
Hm.. it didn't change anything... it was the only limitation I had.
Perhaps I reboot the device again? And we do a last test... As far as I understand, it seems to be an issue of my phone and my conenction and not TS...

**Comment by moxie0** [comment]
@WalterFaber The other possibility is that your network is filtering GCM traffic. Are you on wifi? Have you tried mobile data as well?

The final thing to try is re-registering (uncheck TextSecure Settings -> Advanced -> TextSecure messages, and then re-check it).

**Comment by WalterFaber** [comment]
The device starts now... just a dumb question: is it possible that another app listens to GCM messages and "steals" them (without bad intention). Because there was also an update of a soccer-newsticker-app yesterday... I trust this app, but mybe it absorbes TS GCM messages accidently?

The phone is available now, could you please send a message? 

... Yes, I'm using wifi... no testing with mobile data so far

**Comment by moxie0** [comment]
@WalterFaber Just sent you a message. It's not possible that another app can mess with your GCM state on a non-rooted device. Let's try mobile data next.

**Comment by WalterFaber** [comment]
No, didn't work. The message arrived after I unlocked the database.
Now I turned wifi off, the phone is connected via mobile data.

Of course, I changed nothing related to my wifi network and it also did not work in the office where I use a different wifi.
The phone is not rooted

So, we can do the last test...

**Comment by moxie0** [comment]
Message sent. Looks like it didn't work, try reregistering.

**Comment by WalterFaber** [comment]
That means I get a new "certificate" and my contacts receive a warning message, right?

So if you want me to reregister now I will do it now, because I want to help you to improve the app... but if you don't care because it looks like my local issue, I would wait one or two days... maybe something at google hangs and something/someone fixes it

**Comment by moxie0** [comment]
No as long as you don't uninstall your contacts won't get a warning.  This doesn't change anything to do with encryption.

**Comment by WalterFaber** [comment]
reregistration done... let's see what happens in 1 minute...

**Comment by moxie0** [comment]
Message sent. Looks like it didn't go through.  If that didn't work, there is *definitely* something wrong with your device or your network.

**Comment by WalterFaber** [comment]
No, still the same... ok... I will wait some days. If something changes or I find a solution I will post it here. I'm curious if other users report similar things in playstore or here...

Just a side note:
Today, there was an article in a magazine about network issues which affect Threema users.
I use both carriers mentioned in the article. However, in the office, we use another carrier (which is also mentioned in a remark in the article). I can't imagine this is the same issue, but maybe someone sees a relation...
http://www.heise.de/newsticker/meldung/Krypto-Messenger-Threema-mit-Verbindungsproblemen-2593579.html 

Here's a related link from the article to a comment by Threema. They also posted a tracert... perhaps you find a "known provider" there https://threema.ch/de/blog/posts/verbindungsproblem-o2-und-telefonica-der-grund

Anyway... thank you very much for your kind support! Your app is really great!! It was an honor to do these tests with you. Have a good day and let me know, if you need a tester or any other support.

**Comment by namnamir** [comment]
I have nexus 5 and the same issue. My friends with different phones have the same issue.
All of us use Android 5.0

**Comment by onny** [comment]
I also encountered this problem sometimes with my LG phone (Cyanogenmod installed)

**Comment by moxie0** [comment]
@namnamir @onny If this is a problem that occurs for you on every time the app has been in the background for > 60 seconds, then it is definitely a problem with your device's ability to receive GCM messages.  That usually comes down to some weird xprivacy stuff, disabling background data, a bad gapps install, or a network that filters.

**Comment by McLoo** [comment]
Now me too. 
First I thought it was Sony Stamina mode. But no. 
TS is not greenified or elwise affected. 

https://gist.github.com/ce8294d5d2f4f360157a

**Comment by WalterFaber** [comment]
Here is a last update regarding this issue.
1. IT IS NOT A TEXTSECURE ISSUE - Several other Apps using GCM were affected
2. I suspect that the reason is a (silent) "google play services" update to version 7.0.99 google distributed around end of March 2015
3. In the last 2 years I had many google apps deactivated and everything / GCM worked
4. At the beginning of April 2015, GCM stopped working
5. GCM (and hence Textsecure) worked again, after I activated all these apps again
6. Below I post a list of these apps. In the next days I will try to identify the app which causes the issue. However it is possible that it is not a single app but all of these apps must be activated
7. If you receive Textsecure Messages only after start of the app (or while updating an app via playstore - messages arrived immediately) try to activate all these and other google apps
8. Good luck


List of apps:
chrome
google mail
google search
google+
internet
mail widget
talk
talk back

**Comment by siccovansas** [comment]
I had the same problem on a Fairphone with unrooted Android 4.2.2 after one of the recent updates. The problem disappeared though when I just updated to 2.10.2!

**Comment by McLoo** [comment]
seems fixed for me too with 2.10.2. 
(re-enabling the apps @WalterFaber mentioned didn't work)

**Comment by WalterFaber** [comment]
Hi All,

just another feedback from my side:

At the moment, it worked again. The root cause and the solution is unclear.

Here is what I did in the last week:

1. Starting point: 
a) Messages arrived only after opening TS
b) all these apps were deactivated, all updates were deinstalled
chrome, google mail, google search, google+, internet, mail widget, talk, talk back

2. First approach
a) I activated all these apps at once
b) Messages arrived immediately at any point in time. Even when TS was not running
c) Updates of these apps arrived over several days, I installed all of them

3. Analysis
a) I startet to deactivate one app after another (without deinstallation) in order to isolate the relevant app

4. Findings
Two days ago, I had finally deactivated ALL of these apps again. So even if activating them fixed my issue immediately (until today!), deactivating all of them did not cause the issue again!

So maybe something was hanging or (the still installed) updates delivered a required new file - I don't know, but it works.

Have a good time
WalterFaber
